[숫자 야구 게임 미션] 노명은 미션 제출합니다.

nomyeongeun opened the pull request:
안녕하세요! 늦어서 죄송합니다🥲🥲 오늘 일정이 예정보다 늦어져서 늦었습니다ㅠㅠ

우선 돌아가게 만든 다음 저번 리뷰들 생각하면서 고치다보니 생각보다 시간이 오래 걸렸던 것 같습니다.. 하면서 어려웠던 점들 아래 다시 작성해서 올리겠습니다ㅠㅠ

nomyeongeun committed:
docs : 기능 목록 작성

- 입력, 출력, 게임으로 구분

nomyeongeun committed:
feat : 입력 기능 구현

- 맞출 숫자 입력 받기, 종료 여부 숫자 입력 받기
- 입력 조건에 대한 예외 처리

nomyeongeun committed:
feat : 출력 기능 구현

- 숫자 입력 안내 출력, 게임 결과 출력

nomyeongeun committed:
feat : 게임 실행 부분 구현

nomyeongeun committed:
refactor : 메소드 이름 변경

woosung1223 commented on README.md:
```diff
@@ -134,3 +134,32 @@
+  - 한 게임마다 하나의 새로운 숫자를 생성한다.
+- 입력한 수에 대한 결과를 판단하는 기능
+  - 같은 수가 같은 자리에 있으면 스트라이크
+  - 다른 자리에 있으면 볼
```
기능 목록 💯

기능 목록을 이렇게 작성해보니 어떠셨나요?
기능 목록을 작성하지 않는 방식과 대비해 어떤 장점들을 느끼셨는지 궁금합니다!

코멘트로 남겨주세요~!

nomyeongeun replied:
감사합니다 🙇‍♂️🙇‍♂️ 멘토님께서 저번 강의 때 참고할 예시 보여주신 덕분입니다!!
기능 목록을 입력, 출력, 게임 실행의 세 부분으로 나누고 각각에 대한 기능들을 작성했는데

- 입력, 출력을 게임 실행과 별도로 분리해서 생각해서 좋았던 점
  -   게임 로직에 좀 더 온전하게 집중할 수 있었고, 작성된 코드도 중심이 되는 게임이 어떻게 흘러가는지 눈에 잘 보였습니다
  - 이번에 리뷰 해주신 내용을 바탕으로 코드 수정할 때도 이렇게 분리되어있어서 훨씬 수월했던 것 같습니다!
 
- 각각 기능을 작성할 때는 말씀해주셨던대로 애초부터 좀 더 굵직 굵직한 추상적인 부분부터 쓰고 나서 점점 구체적인 부분으로 세분화해서 기능 목록을 작성해보려고 했었습니다. 이렇게 하니까 나중에 실제로 구현할 때도  애초에 작성된 기능 목록에 맞춰서 처음부터 메소드를 분리해서 구현하니까 (이전에는 쭉 구현하고 나중에 분리했었는데) 편하고 메소드도 더 체계적으로 잘 분리된 것 같다는 느낌이었습니다.
( 게임 실행에서 처음에 숫자를 생성하는 부분과 결과를 판단하는 부분으로 크게 나눠서 기능 목록을 작성했는데, 구현할 때 기능 목록에 맞춰서 숫자 생성하는 부분을 따로 분리해두니까 개인적으로 좋았던 것 같습니다)

woosung1223 replied:
💯

백엔드 개발자가 기능 목록을 작성해야 하는 대부분의 이유를 체득하신 것 같네요!

기능 목록은 구현 가이드라인이 되기도 하면서, 실제로 구현에 들어가기 전에 로직이나 제약조건에 대해 세세하게 생각해볼 수 있는 계기가 되기도 합니다.

이후 Spring Boot를 사용한 프로젝트를 하시게 되더라도 프로덕트 팀과 협의해 기능을 개발해야 하는데, 이런 경우 기능 목록을 작성하고 구현에 뛰어드는게 저는 좋다고 느껴졌습니다!

woosung1223 commented on src/main/java/baseball/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,28 @@
+import camp.nextstep.edu.missionutils.Randoms;
+
+public class RandomNumberGenerator {
+    public int makeThreeDigitNumberMadeOfDifferentNumbers() {
```
클린 코드 수업을 열심히 들으셨군요 👍👍

메소드명은 명은님이 작성해주신 것처럼, 의도를 최대한 드러내는 게 좋습니다.

메소드 바디를 들여다보지 않아도, `3개의 서로 다른 랜덤 숫자를 만드는구나!` 라는 생각이 들겠네요 😄

woosung1223 commented on src/main/java/baseball/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,28 @@
+            third = Randoms.pickNumberInRange(1,9);
+        }
+        return third;
+    }
```
개인적으로 메소드 분리 상당히 깔끔해서 놀랬습니다!..

예전에 제가 해당 미션을 했을 때에는 이렇게 깔끔하게 풀어내지는 못했던 것 같은데, 잘해주셨습니다 💯

woosung1223 commented on src/main/java/baseball/Application.java:
```diff
@@ -1,7 +1,17 @@
     public static void main(String[] args) {
-        //TODO: 숫자 야구 게임 구현
+        while (true) {
+            int answerNumber = new RandomNumberGenerator().makeThreeDigitNumberMadeOfDifferentNumbers();
```
일반적으로 저는 `new 객체().메소드()` 형태보다는,

```java
RandomNumberGenerator randomNumberGenerator = new RandomNumberGenerator();
int answerNumber = randomNumberGenerator.makeThreeDigitNumberMadeOfDifferentNumbers();
```
형태를 선호합니다!

코드 라인이 짧은 건 물론 좋은 일이지만, 
가독성이 개선되는 방향이라면 코드 라인이 조금 늘어나도 상관은 없습니다 😄

woosung1223 commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,56 @@
+public class Game {
+    InputView input = new InputView();
+    OutputView output = new OutputView();
+    private final int answer;
```
좋은 구조입니다 💯

Game이라는 객체가 InputView, OutputView를 활용하고 있네요~

다만 접근제어자만 붙여주시면 더 좋을 듯 합니다!

woosung1223 replied:
그리고 answer를 생성자로 받은 만큼, InputView, OutputView도 생성자로 받을 순 없을까요?

해당 방식과 대비해 생성자로 받는 방식은 어떤 이점이 있을까요?!

nomyeongeun replied:
1차 리뷰에 제가 이해한 내용을 바탕으로 'InputView, OutputView도 생성자로 받도록' 해봤는데, 제가 이해한 바가 맞는지 확실하지 않아서 혹시 틀렸다면 고민해서 다시 고쳐보도록 하겠습니다..🥲

개인적인 생각으로는, 이전보다 입력과 출력 부분이 좀 더 게임 실행 부분과 확실하게 분리된 것 같다는 느낌이 들었습니다. 원래 Game 객체 부분에 있던 InputView와 OutputView 객체를 새로 만드는 부분이  Application으로 넘어가서(밖으로 나와져서) 그렇게 느낀 것 같습니다.

woosung1223 replied:
제가 의도한대로 구현해주신게 맞습니다!

> 개인적인 생각으로는, 이전보다 입력과 출력 부분이 좀 더 게임 실행 부분과 확실하게 분리된 것 같다는 느낌이 들었습니다. 원래 Game 객체 부분에 있던 InputView와 OutputView 객체를 새로 만드는 부분이 Application으로 넘어가서(밖으로 나와져서) 그렇게 느낀 것 같습니다.

사실 이렇게 '설정과 사용'을 분리하는게 결합도가 낮은 설계를 위한 중요한 방법 중 하나입니다. 스프링을 추후에 배우시게 되면, '의존관계 주입' 혹은 'Dependency Injection' 이라는 키워드를 많이 들으실텐데 이번에 명은님이 하신 방법이 이 방법입니다!

어떤 객체를 사용할지는 외부에서 주입받도록 하면 낮은 결합도를 유지할 수 있습니다 👍

woosung1223 commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -1,5 +1,51 @@
-
+import camp.nextstep.edu.missionutils.Console;
 public class InputView {
+    public int userNumber() {
```
일반적으로 메소드명은 동사로 작성합니다.
`~을 한다`에 가깝기 때문입니다!

`readUserNumber`와 같은 이름도 괜찮겠네요!

woosung1223 commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -1,5 +1,51 @@
+    private int checkUserNumberError(String str) {
+        try{
+            int number = Integer.parseInt(str);
+            if (number < 111 || number > 999) {
```
111, 999는 어떤 의미를 갖고 있나요?!

매직 넘버에 대해 한 번 찾아보시길 바랍니다!

woosung1223 commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,56 @@
+    private int calStrike(int number) {
+        int result = 0;
+        char[] answerArray = String.valueOf(answer).toCharArray();
+        char[] userArray = String.valueOf(number).toCharArray();
```
String -> 배열 변환이 필요하지 않은 것 같습니다!

String 내부 메소드를 통해 해결할 수 있는 방법에 대해 찾아보시길 바랍니다 😄

woosung1223 replied:
추가로, 사용자가 입력하는 숫자와 컴퓨터가 생성한 숫자는 모두 3자리인데,

이를 객체로 만들어서 사용해볼 수 있지 않을까요?

한 번 고민해보시길 바랍니다!

woosung1223 reviewed (changes requested):
안녕하세요 명은님!

기능 목록부터 로직까지 아주 잘 작성해 주셨습니다! 저번 미션과 대비해서 또 많이 성장하신 것 같네요.. 💯 

객체 분리에 대한 고민, 클린 코드에 대한 고민을 많이 하셨던 것 같습니다. 코드에서도 드러나네요 👍

워낙 잘하시니까 제가 조금 챌린징한 미션을 추가로 드리도록 하겠습니다.

else if, else문을 완전히 제거해보세요!
메소드 분리를 적절하게 수행하면 대부분의 else문은 제거가 가능합니다.
다만 어려울 수도 있으니 선택적으로 반영해주시면 됩니다!

고생하셨고, 리뷰 반영하시면 또 확인해보도록 할게요~!

nomyeongeun committed:
refactor : 객체 생성과 메소드 호출 분리

- 가독성 개선

nomyeongeun committed:
refactor : Game 객체 생성자, 필드 수정

- InputView, OutputView 변수 접근 제어자
- InputView, OutputView 생성자로 받기

nomyeongeun committed:
refactor : 메소드 이름 수정

- userNumber, exitNumber 메소드 이름이 동사 형태가 되도록 이름 수정

nomyeongeun committed:
fix : 메소드 이름 수정에 맞게 호출 되는 부분도 이름 수정

nomyeongeun committed:
refactor : 불필요한 type 변환 삭제

- Game의 calStrike() : String에서 Character Array로 변환 없이 String 내부 메소드를 통해 같은 동작 하도록 수정

nomyeongeun committed:
refactor : magic number 사용

- InputView의 checkUserNumberError에서 사용자가 입력할 수 있는 최대, 최소값을 상수 변수로 선언해서 의미 나타낸다

nomyeongeun committed:
refactor : Number 객체 추가

- 세자리 수의 값과 각 자릿수를 미리 저장해두는 객체

nomyeongeun committed:
refactor : else if, else 최소화 및 메소드 이름 변경

nomyeongeun commented:
안녕하세요, 1차 리뷰해주신 내용 반영해보았습니다!  이번에는 애매하다 생각했던 부분이나 확신을 못 갖는 부분들이 많았어서, 수정했던 부분 중에서 고민을 많이 했던 부분들에 대해서 남겨보겠습니다.

## ✍🏻 수정 사항과 고민
**1.  else if와 else 없애기**
- OutputView에 있던 else if와 else를 메소드를 분리하지는 않고 if 로 바꿔줬습니다. 결과에 정답, 낫싱, 그 외라는 3가지 경우의 수가 있다고 생각했는데, 이걸 3가지 메소드로 분리하는 게 좋은건지 잘 모르겠다는 생각이 들었습니다. 제가 고친 방식도 else if, else를 없앤 것이지 좋은 방식은 아닌 것 같은데, 어떤 방법도 확신이 안 들었던 것 같습니다.

**2. InputView, OutputView의 접근 제어자 붙이기, 생성자로 받게 수정하기**
- 제가 의도하셨던 바를 제대로 이해했는지 잘 모르겠습니다. 처음에는 생성자에 그냥 InputView, OutputView의 생성을 옮겨뒀었는데, 조금 더 생각해보니까 의도하신 바가 아예 Application 쪽으로 밖으로 빼는 것 같아서 그렇게 수정해서 반영했는데 제가 이해한 게 맞는지 궁금하고, 제가 위에 작성한 이유가 맞는지도 궁금합니다.

**3. magic number**
- 변수 이름에 의미를 담아서 해당 숫자가 갖는 역할을 분명하게 한다는 사용 이유는 이해했습니다! 그런데 제가 했던 것처럼 객체 필드에 private static final int 로 상수로 선언하는 게 맞는 방법인가요? 처음에는 메소드 안에 그냥 일반 변수로 선언했었는데, 조금 더 생각해보니까 바뀌면 안 되는 값이라 상수여야 할 것 같아서 수정하다보니 위치가 이렇게 됐는데 이렇게 하는 방식이 맞는 것인지 궁금합니다.

**4. String -> charArray 변환 없애고 String 내부 메소드 활용하기**
- String.charAt() 사용하는 걸로 수정했는데 의도하셨던 방법이 맞는지 궁금합니다. 볼이랑 스트라이크 계산하는 방법에 대해 고민을 많이 했었는데, 지금 방법도 뭔가 이상하다는 생각이 많이 듭니다.

**5. 숫자에 대한 객체 만들기**
- 처음에는 이렇게 하면 도대체 뭐가 좋은지 잘 모르겠어서 감이 안 잡혀서 한참 고민했었습니다. 고민 끝에 겨우 내린 결론은 볼과 스트라이크를 계산할 때 자릿수를 이름이 잘 지어진 숫자 객체의 메소드로 얻으면 더 이해가 쉬울 것 같다였습니다. 이 객체를 더 잘 활용할 수 있는 방법이 있을까요... 계속 고민해보겠습니다.


## 💭 과제 하면서 고민했던 부분, 느낀점
- 메소드 분리를 이제 "잘" 해야겠다는 생각이 들어서, 오히려 지난 과제보다 메소드 분리하는데 시간이 더 오래걸렸던 것 같습니다. 1주차 리뷰 내용이나 강의 내용도 반영하느라 고민하는 시간이 더 길었던 것 같습니다..🥲 광화문에서 길 가는 사람 붙잡고 물어봐도 이해해야한다는 말씀이 기억에 남아서 최대한 그렇게 하려고 노력을 많이 했고, 고민도 많이 하다가, 지각했네요.. 죄송합니다🙇‍♂️
- 볼과 스트라이크 계산하는 부분 정말 무식하고 직관적이고 제가 사람마냥 비교해서 계산하는 방법처럼 결국 해버렸지만, 고민을 정말 많이 했습니다..

처음 제출할 때 지각 이슈로 글을 못 적었어서 1차 리뷰 반영까지 하고 나니 글이 많이 길어졌네요.. 죄송합니다.. 읽어주셔서 감사합니다.!

woosung1223 reviewed (changes requested):
안녕하세요 명은님! 1차 피드백 반영 고생하셨습니다 💯

고민사항들을 나열해주셨는데, 제 개인적인 의견을 담아보도록 하겠습니다.

> 1. else if 와 else 없애기

if - else 문은 분기문입니다. 분기문은 사실 다음과 같은 함수 호출로 변경이 가능합니다!
```java
public void 무슨_함수() {
    if (뭐시기) {
        다른_함수1();
        return;
    }
    다른_함수2();
```
이런 메소드 구조를 작성하다보면, 대부분의 else문은 제거가 가능합니다.
하지만 이게 '실리가 있는 방법인가'를 따져보면 또 다른 이야긴데, 그냥 if문을 작성하는게 더 좋은 방법일 수도 있습니다. 하지만 제가 이 방법을 제안드린 이유는 메소드 분리를 연습해보셨으면 하는 바램이었기 때문입니다!

가령 `printGameResult`는 다음과 같은 방식으로 else 문 제거가 가능하겠네요!
```java
public void printGameResult(int ball, int strike) {
    if (strike == 3) {
        ...
        return;
    }
    printWhenNotGameOver();
}

public void printWhenNotGameOver() {
    if (ball > 0 && strike > 0) {
        ...
        return;
    }
    printWhenNothing();
}
```

> 2. InputView, OutputView의 접근 제어자 붙이기, 생성자로 받게 수정하기

잘 해주신게 맞습니다! 이 부분은 앞서 코멘트로 달았으므로 추가적인 설명은 생략하도록 하겠습니다. 다만 참고하시면 좋을만한 레퍼런스 하나 첨부드립니다.

https://tecoble.techcourse.co.kr/post/2021-04-27-dependency-injection/

> 3. magic number

맞는 방법입니다! 상수는 `private static final 타입`과 같은 형태로 주로 선언합니다 👍

> 4. String -> charArray 변환 없애고 String 내부 메소드 활용하기

제가 의도한 바가 맞습니다! 다만 이는 5번 고민과 연계되는 답변이긴 한데, Number 객체를 잘 사용하면  Number 객체에게 비교하는 책임을 넘겨줄 수 있지 않을까요? 가령 다음과 같이요!

```java
public class Number {

    ...
    public int countStrike(Number other) {
        ...
    }
```

> 5. 숫자에 대한 객체 만들기

이 부분은 제가 4번에서 답변드린 내용을 고민해보시면 충분할 것 같습니다!

볼과 스트라이크 계산은 if문 순수 비교가 아니라 반복문으로도 해결할 수 있을 것 같습니다!
잘 생각해보면, 볼은 `같은 자리의 수는 아니지만 해당 수를 포함하는 경우`이고, 스트라이크는 `같은 자리의 수`인 경우죠!

반복문으로 해결 가능하다면 정책이 바뀌어서 숫자가 4개, 5개, 10개가 되었을때도 쉽게 대응이 가능할 것 같습니다 💯

nomyeongeun committed:
refactor : 분기문을 메소드 분리로 변경

loveysuby commented:
안녕하세요! Java를 처음 써보셨다는 게 진짜 놀라울만큼 구현을 잘 하신 것 같습니다.
멘토님께서 상당히 많은 포인트를 짚어주셔서 마음 편하게 리뷰하겠습니다. 😄

loveysuby commented on src/main/java/baseball/Number.java:
```diff
@@ -0,0 +1,31 @@
+        return number;
+    }
+
+    public int getFirstDigit() {
```
```java
if (( answer.getThirdDigit() == userNumber.getFirstDigit() ) 
         || ( answer.getThirdDigit() == userNumber.getSecondDigit() )) {
    result++;
}
``` 
자리를 끊어서 숫자 간 판단을 하도록 구현하신 부분은 저는 생각하지 못했던 부분인데, 숫자 판단을 할 때 코드가 깔끔해지네요! 

만약 `판단해야 하는 숫자가 5자리로 늘어나거나`, 사용자가 숫자를 입력하는 만큼 컴퓨터가 랜덤 숫자를 생성해야 되는 상황으로 요구사항이 확장되는 경우엔 어떻게 최적화할 수 있을까요 ? 본 미션에 국한된다면 `Over thinking` 일 수 있지만, 숫자 야구 `게임` 을 만들고 계신다면 이런 고민은 설계 단계부터 고민해봐야 하는 문제라고 생각합니다.

Collection 을 활용해서 개선해볼 수 있을 것 같은데, 이런 방법도 있다는 걸 알아두시면 도움이 되실 것 같아요.

nomyeongeun replied:
안녕하세요! 리뷰 해주셔서 정말 감사합니다🙇‍♂️🙇‍♂️
이번 과제 하면서 가장 오래 고민했던 부분이 이 볼과 스트라이크 계산하는 부분이었던 것 같아요!

처음에는 반복분을 사용하는 것보다, 저렇게 자리를 끊어서 판단하는게 더 직관적이라고 생각했던 것 같아서 그렇게 구현했던 것 같아요! 인간인 제가 볼과 스트라이크를 계산했던 절차를 생각해서 했었습니다! 코드도 깔끔해보였구요..!

그런데 멘토님과 효섭님 리뷰도 받아보면서, 3자리 이상의 숫자로 확장될 경우에 대해 더 고민해보게 되었습니다! 저도 처음에 아예 생각해보지 않았던 건 아니지만, 코드를 봤을 때 이해하기 쉽거나 깔끔해지는 부분과 이 부분을 저울질했던 것 같아요.

저를 포함해서 여러 사람이 이 숫자 확장 부분에 대해 고민해본다는 것은, 효섭님 말씀처럼 설계 단계에서 유지보수를 했을 때 고려해야 하는 부분인 것 같다는 판단이 들었습니다! 지금 방법에서도 고쳐야 할 부분이 많겠지만, 그래도 자릿수 확장하기 쉬운 쪽으로 수정해봤습니다! 감사합니다🙏

loveysuby commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -1,5 +1,32 @@
+import java.util.StringJoiner;
+
 public class OutputView {
+    public void printGuessNumberInstruction() {
```
`StringJoiner` 에서 사용되고 있는 `"볼"` , `"스트라이크"` , `"낫싱"` 같은 value 들은 field로 사전에 정의해두고 쓰면 출력 메소드를 디버깅할 때 필드만 보면 되므로, 같은 기능을 하더라도 가독성을 향상시킬 수 있을 것 같아요.

nomyeongeun replied:
효섭님 의견 반영해서 수정해봤는데, 의도하신 걸 제가 잘 이해했을까요? 

이번에 리뷰해신 부분처럼 필드에 사전 정의해두면 좋을 value들에 대해 늘 고민해보는 습관 가져야겠어요! 감사합니다 👍🏻

loveysuby commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -1,5 +1,53 @@
+
+    private int checkUserNumberError(String str) {
+        try{
+            int number = Integer.parseInt(str);
```
검증 로직에서 아래와 같은 중복된 변환 로직이 존재하는데, 혹시 어떤 이유에서 이렇게 하셨는지 이유를 알 수 있을까요?

```java
int number = Integer.parseInt(str);
```

nomyeongeun replied:
제 한국어 독해 능지 이슈로 이해를 못 했는데...🥲 효섭님 리뷰 받고 검증 부분을 많이 수정해서 관련된 부분은 아래에서 자세히 작성할게요!

말씀해주신 부분에 대해서 중복된 변환 로직이 존재한다는게 어떤 의미인지 조금 더 자세히 설명해주실 수 있으실까요.. ?

loveysuby replied:
검증 메소드에서 `Console.readLine()`에 의해 String 입력값을 받고 있는데, 개인적으로 이 부분은 메소드가 검증하는 대상을 어떻게 정의할 것인지에 따라 차이가 존재한다고 생각해요. 성능에 영향을 주지는 않는 부분입니다.

검증 메소드는 Number를 다루고 있네요. Integer 클래스로 변환하는 부분을 매번 검증 시마다 선언하는 방법 대신  아예 Integer number를 입력받고, exitNumber를 Integer로 미리 바꿔준 다음에 숫자에 대해서 검증할 수도 있겠네요.

`Integer.parseInt(str);` 로 매 번 변환되는 코드를 줄이고, 검증하려는 대상 역시 `checkNumeric`이 통과하면 굳이 `checkUsreNumberError`에서 `String`으로 받지 않아도 같은 결과를 만들어낼 수 있을 것 같아요.

loveysuby commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,58 @@
+    private final InputView input;
+    private final OutputView output;
+    private final Number answer;
+
```
`Game` controller가 `View`, `Number` 를 의존하도록 구현하셨네요!
뷰에 관련된 로직은 Number에 대한 `getFirstDigit`의 로직을 알지 못하게 되네요. 객체들에 부여된 책임이 `Game` 객체 덕분에 분명해지는 것 같습니다.

loveysuby commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -1,5 +1,53 @@
+        int number = 0;
+        try {
+            number = checkUserNumberError(str);
+        } catch (IllegalArgumentException e) {
```
`checkUserNumberError` 메소드는 Client가 입력한 숫자가 요구사항의 기준을 지키고 있는지 검증하고 있네요. 입력과 검증을 분리하려는 관점이 탁월하신 것 같습니다!

검증에 실패하면 try-catch에 의해 `입력값을 읽는 메소드` 에서 `IllegalArgumentException`를 발생시키고 있네요.

검증에 실패한 이후의 행동까지 검증 메소드에서 수행될 수도 있을 것 같은데, 이 부분을 어떻게 위임할지 고민해보시면 입력과 검증을 정확히 분리할 수 있을 것 같아요!

nomyeongeun replied:
너무 좋은 말씀이신 것 같아요 👍🏻👍🏻👍🏻👍🏻👍🏻 말씀해주신대로 입력과 검증 부분을 좀 더 정확히 분리해봤습니다..! 

숫자가 아닌 값이 들어온 경우 `Integer.parseInt()`에서 발생할 수 있는 에러를 `checkNumeric`으로 해서 입력받는 두 경우 모두 확인하게 하고, 나머지 맞출 숫자나 종료 숫자에 대한 검증을 각각 만들어뒀습니다! 검증 부분에서도 반환을 안 하는 쪽으로 수정해봤어요!

훨씬 더 좋은 구조가 된 것 같아요! 감사합니다~!~!

nomyeongeun committed:
refactor : Number에서 strike와 ball 계산

nomyeongeun committed:
refactor : 입력값 검증 부분 수정

nomyeongeun committed:
refactor : 볼, 스트라이크, 낫싱 메세지 필드로 지정

woosung1223 reviewed (approved):
안녕하세요 명은님!

확인해봤는데 효섭님 리뷰까지 제대로 반영해주셨네요 👍

효섭님이 말씀해주신대로 자바를 처음 사용하는 것이라고는 잘 생각이 들지 않는 것 같네요 😀

다음 미션 진행해주세요! 고생하셨습니다!
